test/end2end_test.go

```diff
@@ -7668,3 +7668,82 @@ func (s) TestClientSettingsFloodCloseConn(t *testing.T) {
 	s.GracefulStop()
 	timer.Stop()
 }
+
+// TestDeadlineSetOnConnectionOnClientCredentialHandshake tests that there is a deadline
+// set on the net.Conn when a credential handshake happens in http2_client.
+func (s) TestDeadlineSetOnConnectionOnClientCredentialHandshake(t *testing.T) {
+	lis, err := net.Listen("tcp", "localhost:50053")
+	if err != nil {
+		t.Fatalf("Failed to listen: %v", err)
+	}
+	go func() {
+		lis.Accept()
+	}()
+	credsInvoked := testutils.NewChannel()
+	cvd := &credentialsVerifyDeadline{
+		credsInvoked: credsInvoked,
+	}
+	dopts := []grpc.DialOption{
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			conn, err := (&net.Dialer{}).DialContext(ctx, "tcp", addr)
+			if err != nil {
+				return nil, err
+			}
+			cpd := &connPersistDeadline{
+				Conn: conn,
+			}
+			return cpd, nil
+		}),
+		grpc.WithTransportCredentials(cvd),
+	}
+	cc, err := grpc.Dial(lis.Addr().String(), dopts...)
+	if err != nil {
+		t.Fatalf("Failed to dial: %v", err)
+	}
+	defer cc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	connectionHasCorrectDeadline, err := credsInvoked.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Error receiving from credsInvoked: %v", err)
+	}
+	if !connectionHasCorrectDeadline.(bool) {
+		t.Fatalf("Connection did not have deadline set.")
+	}
+}
+
+type connPersistDeadline struct {
+	net.Conn
+	deadline time.Time
+}
+
+func (c *connPersistDeadline) SetDeadline(t time.Time) error {
+	c.deadline = t
+	return c.Conn.SetDeadline(t)
+}
+
+type credentialsVerifyDeadline struct {
+	credsInvoked *testutils.Channel
```

Comment:
How about just embedding a `credentials.TransportCredentials` and initialize with a `credentials/local.NewCredentials()`.  Then you only need to override `ClientHandshake`.

I'd make a constructor for this, too.

Reply:
I like it like this. A. it keeps it with consistent with the other fake credential in the test. B. I don't like the idea of tying this to an actual credential implementation, as the net.Conn actually had functionality that needed to happen, whereas the sole singular purpose of these credentials is to take the deadline off the conn and pass it on a channel.